File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    
    visited= [False]*(len(dungeons))
    
    # k:현재 피로도
    # 최소>= 소모 
       
    answer = 0
    def dfs(cur_hp,step):
        nonlocal answer
        answer= max(step,answer)
        
        
        for i in range(0,len(dungeons)):
            min_hp,lose= dungeons[i]
            if not visited[i] and cur_hp>= min_hp :
                visited[i]= True
                dfs(cur_hp-lose, step+1)
                visited[i]= False

    dfs(k,0)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
class Solution {
    int answer =0;
    boolean[] visited;
    public int solution(int k, int[][] dungeons) {
        
        visited= new boolean [dungeons.length];
        
        dfs (k,dungeons,0);
        
        return answer;
    }
    
    public void dfs (int cur_hp, int[][] dungeons, int cnt) {
        answer = Math.max(cnt, answer);
        
        for (int i=0; i< dungeons.length; i++) {
            int need = dungeons[i][0];
            int cost = dungeons[i][1];
            
            if (!visited[i] && cur_hp >= need) {
                visited[i] = true;
                dfs (cur_hp - cost, dungeons, cnt+1);
                visited[i] = false;
            }
         
            
        }
    }
}